### PR TITLE
Languages

### PR DESCRIPTION
**Pull request by ikonitas**

### CODE REVIEW

File: glitter/pages/models.py [hunk 1]
```diff
@@ -23,27 +27,63 @@ def unpublished(self):
...
+    @staticmethod
+    def is_login_required():
+        if hasattr(settings, 'GLITTER_SHOW_LOGIN_REQUIRED'):
+            return getattr(settings, 'GLITTER_SHOW_LOGIN_REQUIRED')
+        return False
+
     def get_absolute_url(self):
-        return self.url
+        url = self.url
+        if hasattr(settings, 'PAGE_LANGUAGES'):
```
Comment (tomkins) [inline glitter/pages/models.py]: Think this should be `if self.language`?

File: glitter/pages/admin.py [hunk 2]
```diff
@@ -20,27 +20,89 @@
...
+    def get_list_display(self, request):
+        """ Add languages to the list_display if it's enabled. """
+
+        if self.model.is_languages_required():
+            if 'get_language' not in self.list_display:
+                self.list_display.append('get_language')
+        return self.list_display
+
+    def get_url(self, obj):
+        url = obj.url
```
Comment (tomkins) [inline glitter/pages/admin.py]: Just use `obj.get_absolute_url()` - less duplication.

File: glitter/pages/admin.py [hunk 3]
```diff
@@ -20,27 +20,89 @@
...
+
     def view_url(self, obj):
         info = self.model._meta.app_label, self.model._meta.model_name
         redirect_url = reverse('admin:%s_%s_redirect' % info, kwargs={'object_id': obj.id})
         return '<a href="%s">View page</a>' % (redirect_url)
     view_url.short_description = 'View page'
     view_url.allow_tags = True
 
+    def get_language(self, obj):
+        return dict(settings.PAGE_LANGUAGES)[obj.language].title()
```
Comment (tomkins) [inline glitter/pages/admin.py]: This fails if `obj.language` is an empty string.

File: glitter/pages/middleware.py [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
 from django.http import Http404, HttpResponseRedirect
 
 from glitter.exceptions import GlitterRedirectException, GlitterUnpublishedException
+from glitter.pages.views import glitter, render_object_unpublished
 
 
 class PageFallbackMiddleware(object):
     def process_response(self, request, response):
 
         # Import here as it causes migrations for pages if app is not installed apps.
```
Comment (tomkins) [inline glitter/pages/middleware.py]: The comment no longer matches the code?

**Comment by zatan** [comment]
RAaaddhy

**Comment by tomkins** [comment]
On hold - as we'll need to do more work with this for navigation template tags... which also needs additional work/effort.

**Comment by tomkins** [comment]
As mentioned earlier - let's revisit this at a later date.

I want languages/navigation to work well with each other and be first class features of glitter. We can possibly take a few new tricks from more recent projects.

**Pull request closed without merging**